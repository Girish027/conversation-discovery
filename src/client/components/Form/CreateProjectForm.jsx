import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  TextField, Download, Doc, Button
} from '@tfs/ui-components';
import constants from '../../constants';
import Dropzone from './Dropzone';
import CsvIcon from '../Icons/CsvIcon';
import { shouldHandle } from 'utils/KeyboardUtils';
import { downloadFile } from 'utils/FileUtils';

const ERROR_TYPES = {
  DUPLICATE_PROJECT_NAME: 'DUPLICATE_PROJECT_NAME',
  INVALID_FILE_COLUMNS: 'INVALID_FILE_COLUMNS',
};

export default class CreateProjectForm extends PureComponent {
  constructor(props) {
    super(props);
    this.props = props;
    this.onChange = this.onChange.bind(this);

    this.onDownloadKeyPress = this.onDownloadKeyPress.bind(this);
    this.downloadDatasetTemplate = this.downloadDatasetTemplate.bind(this);
  }

  onChange(event) {
    const { value, name } = event.target;
    const { saveChanges } = this.props;
    saveChanges({ [name]: value });
  }

  onDownloadKeyPress(event) {
    if (shouldHandle(event)) {
      this.downloadDatasetTemplate();
    }
    event.preventDefault();
  }

  downloadDatasetTemplate() {
    const { caa } = this.props;
    const locationUrl = constants.serverApiUrls.downloadTemplate(caa);
    downloadFile(locationUrl);
  }

  render() {
    const {
      error, errorType, saveFile, formdata
    } = this.props;

    const { projectName = '', projectDescription = '' } = formdata;
    return (
      <div>
        <div className='formItem'>
          <div className='labelName'>Name</div>
          <TextField
            type='text'
            name='projectName'
            onChange={this.onChange}
            styleOverride={{
              width: '390px',
            }}
            defaultValue={projectName}
          />
        </div>
        <div className='formItem'>
          <div className='labelName'>Description</div>
          <TextField
            type='text'
            name='projectDescription'
            onChange={this.onChange}
            styleOverride={{
              width: '390px',
            }}
            defaultValue={projectDescription}
          />
        </div>
        <div className='formItem'>
          <div className='labelName'>
            <span> Upload Chat Data File </span>
            <hr />
          </div>
          <Button
            onClick={this.downloadDatasetTemplate}
            onKeyDown={this.onDownloadKeyDown}
            type='flat'
            styleOverride={{
              textDecoration: 'underline',
              fontSize: '12px',
              color: '#313f54',
              fontWeight: 'normal'
            }}
          >
            <Download
              width={15}
              height={15}
              fill='#004c97'
              style={{
                verticalAlign: 'middle',
                paddingRight: '10px',
              }}
            />
            Download the template
          </Button>
          <Dropzone
            accept={'text/csv, application/vnd.ms-excel'}
            icon={CsvIcon}
            acceptedIcon={Doc}
            multiple={false}
            saveFile={saveFile}
            maxSize={constants.maxDatasetFileSize}
          />
          { errorType === ERROR_TYPES.INVALID_FILE_COLUMNS && (
            <span className='warning' style={{ height: '15px' }}>
              { error || null }
            </span>
          )}
        </div>
      </div>
    );
  }
}

CreateProjectForm.propTypes = {
  saveFile: PropTypes.func,
  saveChanges: PropTypes.func,
  error: PropTypes.string,
  errorType: PropTypes.string,
  caa: PropTypes.object.isRequired,
  formdata: PropTypes.object,
};

CreateProjectForm.defaultProps = {
  saveFile: () => {},
  saveChanges: () => {},
  error: '',
  errorType: '',
  formdata: {},
};
