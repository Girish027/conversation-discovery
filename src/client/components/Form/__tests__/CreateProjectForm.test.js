import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import CreateProjectForm from 'components/Form/CreateProjectForm';

describe('<CreateProjectForm />', function () {
  let wrapper;


  const props = {
    caa: {
      clientId: 'a',
      accountId: 'b',
      appid: 'c'
    },
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  const getShallowWrapper = (propsObj) => shallow(<CreateProjectForm
    {...propsObj}
  />);

  const getMountedWrapper = (propsObj) => mount(<CreateProjectForm
    {...propsObj}
  />);

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getMountedWrapper(props);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly with default props', () => {
      wrapper = getShallowWrapper(props);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
    test('renders correctly with formdata when provided', () => {
      wrapper = getShallowWrapper({
        ...props,
        formdata: {
          projectName: 'pro-1',
          projectDescription: 'desc'
        }
      });
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality', function () {
    beforeEach(function () {
      props.saveChanges = jest.fn();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    describe('onChange', function () {
      test('should save the changes to value in the form', () => {
        wrapper = getShallowWrapper(props);
        wrapper.instance().onChange({
          target: {
            name: 'projectName',
            value: 'new proj name'
          }
        });
        expect(props.saveChanges).toHaveBeenCalledWith({
          projectName: 'new proj name'
        });
      });
    });

    describe('onDownloadKeyPress', function () {
      test('should download when Enter is pressed ', () => {
        wrapper = getShallowWrapper(props);
        wrapper.instance().downloadDatasetTemplate = jest.fn();
        wrapper.instance().onDownloadKeyPress({
          keyCode: 13,
          preventDefault: () => {}
        });
        expect(wrapper.instance().downloadDatasetTemplate).toHaveBeenCalled();
      });

      test('should download when Enter is pressed ', () => {
        wrapper = getShallowWrapper(props);
        wrapper.instance().downloadDatasetTemplate = jest.fn();
        wrapper.instance().onDownloadKeyPress({
          keyCode: 14,
          preventDefault: () => {}
        });
        expect(wrapper.instance().downloadDatasetTemplate).not.toHaveBeenCalled();
      });
    });
  });
});
