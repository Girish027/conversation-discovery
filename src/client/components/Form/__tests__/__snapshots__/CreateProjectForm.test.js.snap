// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProjectForm /> Snapshots renders correctly with default props 1`] = `
<div>
  <div
    className="formItem"
  >
    <div
      className="labelName"
    >
      Name
    </div>
    <withTheme(TextField)
      defaultValue=""
      name="projectName"
      onChange={[Function]}
      styleOverride={
        Object {
          "width": "390px",
        }
      }
      type="text"
    />
  </div>
  <div
    className="formItem"
  >
    <div
      className="labelName"
    >
      Description
    </div>
    <withTheme(TextField)
      defaultValue=""
      name="projectDescription"
      onChange={[Function]}
      styleOverride={
        Object {
          "width": "390px",
        }
      }
      type="text"
    />
  </div>
  <div
    className="formItem"
  >
    <div
      className="labelName"
    >
      <span>
         Upload Chat Data File 
      </span>
      <hr />
    </div>
    <withTheme(Button)
      onClick={[Function]}
      styleOverride={
        Object {
          "color": "#313f54",
          "fontSize": "12px",
          "fontWeight": "normal",
          "textDecoration": "underline",
        }
      }
      type="flat"
    >
      <Download
        fill="#004c97"
        height={15}
        style={
          Object {
            "paddingRight": "10px",
            "verticalAlign": "middle",
          }
        }
        width={15}
      />
      Download the template
    </withTheme(Button)>
    <Dropzone
      accept="text/csv, application/vnd.ms-excel"
      acceptStyle={
        Object {
          "borderColor": "#00e676",
        }
      }
      acceptedIcon={[Function]}
      activeStyle={
        Object {
          "borderColor": "#2196f3",
        }
      }
      baseStyle={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fafafa",
          "borderColor": "#eeeeee",
          "borderRadius": 2,
          "borderStyle": "dashed",
          "borderWidth": 2,
          "color": "#bdbdbd",
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": "90px",
          "outline": "none",
          "padding": "20px",
          "transition": "border .24s ease-in-out",
        }
      }
      disabled={false}
      icon={[Function]}
      maxSize={200000000}
      multiple={false}
      rejectStyle={
        Object {
          "borderColor": "#ff1744",
        }
      }
      saveFile={[Function]}
      shouldUpload={false}
      uploadHandler={[Function]}
    />
  </div>
</div>
`;

exports[`<CreateProjectForm /> Snapshots renders correctly with formdata when provided 1`] = `
<div>
  <div
    className="formItem"
  >
    <div
      className="labelName"
    >
      Name
    </div>
    <withTheme(TextField)
      defaultValue="pro-1"
      name="projectName"
      onChange={[Function]}
      styleOverride={
        Object {
          "width": "390px",
        }
      }
      type="text"
    />
  </div>
  <div
    className="formItem"
  >
    <div
      className="labelName"
    >
      Description
    </div>
    <withTheme(TextField)
      defaultValue="desc"
      name="projectDescription"
      onChange={[Function]}
      styleOverride={
        Object {
          "width": "390px",
        }
      }
      type="text"
    />
  </div>
  <div
    className="formItem"
  >
    <div
      className="labelName"
    >
      <span>
         Upload Chat Data File 
      </span>
      <hr />
    </div>
    <withTheme(Button)
      onClick={[Function]}
      styleOverride={
        Object {
          "color": "#313f54",
          "fontSize": "12px",
          "fontWeight": "normal",
          "textDecoration": "underline",
        }
      }
      type="flat"
    >
      <Download
        fill="#004c97"
        height={15}
        style={
          Object {
            "paddingRight": "10px",
            "verticalAlign": "middle",
          }
        }
        width={15}
      />
      Download the template
    </withTheme(Button)>
    <Dropzone
      accept="text/csv, application/vnd.ms-excel"
      acceptStyle={
        Object {
          "borderColor": "#00e676",
        }
      }
      acceptedIcon={[Function]}
      activeStyle={
        Object {
          "borderColor": "#2196f3",
        }
      }
      baseStyle={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fafafa",
          "borderColor": "#eeeeee",
          "borderRadius": 2,
          "borderStyle": "dashed",
          "borderWidth": 2,
          "color": "#bdbdbd",
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "height": "90px",
          "outline": "none",
          "padding": "20px",
          "transition": "border .24s ease-in-out",
        }
      }
      disabled={false}
      icon={[Function]}
      maxSize={200000000}
      multiple={false}
      rejectStyle={
        Object {
          "borderColor": "#ff1744",
        }
      }
      saveFile={[Function]}
      shouldUpload={false}
      uploadHandler={[Function]}
    />
  </div>
</div>
`;
